cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/existingdataproc/DataprocClient.java

```diff
@@ -0,0 +1,294 @@
+ */
+final class DataprocClient implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DataprocClient.class);
+  // something like 2018-04-16T12:09:03.943-07:00
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'hh:mm:ss.SSSX");
+  static final String DATAPROC_GOOGLEAPIS_COM_443 = "-dataproc.googleapis.com:443";
+
+  private final ExistingDataprocConf conf;
+  private final ClusterControllerClient client;
+  private final Compute compute;
+  private final String projectId;
+
+  private enum PeeringState {
+    ACTIVE,
+    INACTIVE,
+    NONE
+  }
+
+  static DataprocClient fromConf(ExistingDataprocConf conf) throws IOException, GeneralSecurityException {
+    ClusterControllerClient client = getClusterControllerClient(conf);
+    Compute compute = getCompute(conf);
+    return new DataprocClient(new ExistingDataprocConf(conf), client, compute);
+  }
+
+
+  /**
+   * Extracts and returns the zone name from the given full zone URI.
+   */
+  private static String getZone(String zoneUri) {
+    int idx = zoneUri.lastIndexOf("/");
+    if (idx <= 0) {
+      throw new IllegalArgumentException("Invalid zone uri " + zoneUri);
+    }
+    return zoneUri.substring(idx + 1);
+  }
+
+  private static ClusterControllerClient getClusterControllerClient(ExistingDataprocConf conf) throws IOException {
+    CredentialsProvider credentialsProvider = FixedCredentialsProvider.create(conf.getDataprocCredentials());
+
+    String regionalEndpoint = conf.getRegion() + DATAPROC_GOOGLEAPIS_COM_443;
+
+    ClusterControllerSettings controllerSettings = ClusterControllerSettings.newBuilder()
+      .setCredentialsProvider(credentialsProvider)
+      .setEndpoint(regionalEndpoint)
+      .build();
+    return ClusterControllerClient.create(controllerSettings);
+  }
+
+  private static Compute getCompute(ExistingDataprocConf conf) throws GeneralSecurityException, IOException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    return new Compute.Builder(httpTransport, JacksonFactory.getDefaultInstance(), conf.getComputeCredential())
+      .setApplicationName("cdap")
+      .build();
+  }
+
+  private DataprocClient(ExistingDataprocConf conf, ClusterControllerClient client, Compute compute) {
+    this.projectId = conf.getProjectId();
+    this.conf = conf;
+    this.client = client;
+    this.compute = compute;
+  }
+
+  /**
+   * Update the Label on Dataproc cluster to indicate usage of CDF. This will be used for later
+   * tracking on usage of Dataproc by CDF.
+   *
+   * @param labels Key/Value pairs to set on the Dataproc cluster.
+   * @param name   the cluster name
+   */
+  public ClusterOperationMetadata setSystemLabels(Map<String, String> labels, String name) throws Exception {
+    try {
+      Optional<Cluster> clusterOptional = getDataprocCluster(name);
+      Cluster cluster = clusterOptional.get();
```

Comment:
What if the Optional is empty?

Reply:
Checking if present else throwing an exception.